common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -78,6 +81,27 @@ public int hashCode() {
     return Objects.hash(months, days, microseconds);
   }
 
+  @Override
+  public int compare(CalendarInterval that) {
+    long thisAdjustDays = this.microseconds / MICROS_PER_DAY + this.days + this.months * 30;
```

Comment:
@MaxGekk IIRC we have a different definition of days-per-month in Spark?

Reply:
https://github.com/apache/spark/blob/a4382f7fe1c36a51c64f460c6cb91e93470e0825/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L38
I just find one

Reply:
Only 31 days per months is used at the moment, see https://github.com/apache/spark/blob/a4382f7fe1c36a51c64f460c6cb91e93470e0825/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L333-L343

Reply:
30 is for PostgreSQL compatibility in https://github.com/apache/spark/blob/a4382f7fe1c36a51c64f460c6cb91e93470e0825/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L92-L98

Reply:
ok let's use 30 here

Reply:
let's use `DAYS_PER_MONTH` instead of 30

Reply:
shall we define all these interval constants in one place, like `org.apache.spark.unsafe.types.IntervalConstants`?

Reply:
How about to extract the code:
```scala
   var result = interval.microseconds 
   result += DateTimeUtils.MICROS_PER_DAY * interval.days 
   result += MICROS_PER_YEAR * (interval.months / MONTHS_PER_YEAR) 
   result += MICROS_PER_MONTH * (interval.months % MONTHS_PER_YEAR)
```
and reuse it in the comparison? At least, we will have consistent behavior.

Reply:
> How about to extract the code:
> 
> ```scala
>    var result = interval.microseconds 
>    result += DateTimeUtils.MICROS_PER_DAY * interval.days 
>    result += MICROS_PER_YEAR * (interval.months / MONTHS_PER_YEAR) 
>    result += MICROS_PER_MONTH * (interval.months % MONTHS_PER_YEAR)
> ```
> 
> and reuse it in the comparison? At least, we will have consistent behavior.

It not accurate enough for MICROS_PER_YEAR

Reply:
Your code is not accurate too. 30 * 12 = 360 is slightly far away from average days per year - 365.2425, see https://github.com/apache/spark/pull/25998

Reply:
```
postgres=# select interval '1 year' = interval '360 days';
 ?column?
----------
 t
(1 row)

postgres=# select interval '1 mon' = interval '30 days';
 ?column?
----------
 t
(1 row)

postgres=# select interval '1 mon' = interval '31 days';
 ?column?
----------
 f
(1 row)

postgres=# select interval '1 year' = interval '365 days';
 ?column?
----------
 f
(1 row)
```
If we are more likely to keep compatibility with pg, so a year is 360 days not 365 days, a month  is 30 days

Reply:
In Postgres, interval comparison will treat year as 12 months if comparing with month, 360 days if comparing day and 8640 hours if comparing with hours. This is different from date add/subtract, where '1 year' is based on what kind of year it is added to.

This is same as interval division

```sql
postgres=# select interval '1 year' / 360;
 ?column?
-----------
 1 0:00:00
(1 row)

postgres=# select interval '1 year' / 365;
   ?column?
---------------
 23:40:16.4064
(1 row)
```